contracts/main/CurveTricryptoOptimizedWETH.vy

```diff
@@ -300,117 +299,89 @@ def __init__(
 # ------------------- Token transfers in and out of the AMM ------------------
 
 
-@payable
-@external
-def __default__():
-    if msg.value > 0:
-        assert WETH20 in coins
-
-
 @internal
 def _transfer_in(
-    _coin: address,
+    _coin_idx: uint256,
     dx: uint256,
-    dy: uint256,
-    mvalue: uint256,
-    callbacker: address,
-    callback_sig: bytes32,
     sender: address,
-    receiver: address,
-    use_eth: bool
+    expect_optimistic_transfer: bool,
 ):
     """
     @notice Transfers `_coin` from `sender` to `self` and calls `callback_sig`
             if it is not empty.
-    @dev The callback sig must have the following args:
-         sender: address
-         receiver: address
-         coin: address
-         dx: uint256
-         dy: uint256
-    @params _coin address of the coin to transfer in.
+    @params _coin_idx uint256 Index of the coin to transfer in.
     @params dx amount of `_coin` to transfer into the pool.
-    @params dy amount of `_coin` to transfer out of the pool.
-    @params mvalue msg.value if the transfer is ETH, 0 otherwise.
-    @params callbacker address to call `callback_sig` on.
-    @params callback_sig signature of the callback function.
     @params sender address to transfer `_coin` from.
-    @params receiver address to transfer `_coin` to.
-    @params use_eth True if the transfer is ETH, False otherwise.
+    @params expect_optimistic_transfer bool True if pool expects user to transfer.
+            This is only enabled for exchange_received.
     """
+    received_amounts: uint256 = 0
+    coin_balance: uint256 = ERC20(coins[_coin_idx]).balanceOf(self)
+    recorded_balance: uint256 = self.balances[_coin_idx]
 
-    if use_eth and _coin == WETH20:
-        assert mvalue == dx  # dev: incorrect eth amount
-    else:
-        assert mvalue == 0  # dev: nonzero eth amount
+    # Adjust balances before handling transfers
+    self.balances[_coin_idx] += dx
 
-        if callback_sig == empty(bytes32):
+    # Handle transfers:
+    if expect_optimistic_transfer:
 
-            assert ERC20(_coin).transferFrom(
-                sender, self, dx, default_return_value=True
-            )
+        # Only enabled in exchange_received: it expects the caller
+        # of exchange_received to have sent tokens to the pool before
+        # calling this method.
+        received_amounts = coin_balance - recorded_balance
```

Comment:
So here we need `received_amounts = coin_balance - self.balances[coin_idx]`

Reply:
well we need the recorded balances before updating balances. so this won't do.